[JBTM-3355] harmonize the LRA DTO usage to use only LRAData

ochaloup opened the pull request:
https://issues.redhat.com/browse/JBTM-3355

This is a step before the JBTM-3294 where only one Coordinator LRA DTO object is to be used. The `LRAStatusHolder` is replaced by `LRAData` where is needed and status calls are a bit narrowed.

MAIN LRA
!QA_JTA !QA_JTS_JDKORB !QA_JTS_OPENJDKORB !QA_JTS_JACORB !BLACKTIE !XTS !PERF NO_WIN !RTS !AS_TESTS !TOMCAT !JACOCO

xstefank commented on rts/lra/lra-coordinator-jar/src/main/java/io/narayana/lra/coordinator/domain/model/LRAData.java:
```diff
@@ -97,18 +83,102 @@ public int hashCode() {
+    }
+    public static interface BuilderFinal {
+        LRAData build();
+    }
```
Sorry, but I think that the Builder here is overcomplicated. What would be the reason to keep the order of methods in the builder? Also since all `BuilderWith*` interfaces have only one method which represents next step in the building process then all parameters are required, and so this is just a really complicated way of writing an all args constructor :)

ochaloup replied:
I can remove the `Builder` if you both think it's over-complicated. My position is that the creation of an object with constructor of 8 arguments are pretty messy and builder helps here to put all the values at the right place.

xstefank commented on rts/lra/lra-coordinator-jar/src/main/java/io/narayana/lra/coordinator/domain/model/Transaction.java:
```diff
@@ -145,12 +157,8 @@ public boolean save_state(OutputObjectState os, int ot) {
-                os.packBoolean(true);
-                os.packString(status.name());
-            }
+            os.packBoolean(true);
```
is boolean before status required?

mmusgrov replied:
Agreed, status can never be null.

ochaloup replied:
+1

xstefank commented on rts/lra/lra-coordinator-jar/src/main/java/io/narayana/lra/coordinator/domain/model/Transaction.java:
```diff
@@ -269,7 +277,7 @@ public boolean restore_state(InputObjectState os, int ot) {
             startTime = os.unpackBoolean() ? LocalDateTime.ofInstant(Instant.ofEpochMilli(os.unpackLong()), ZoneOffset.UTC) : null;
             finishTime = os.unpackBoolean() ? LocalDateTime.ofInstant(Instant.ofEpochMilli(os.unpackLong()), ZoneOffset.UTC) : null;
-            status = os.unpackBoolean() ? LRAStatus.valueOf(os.unpackString()) : null;
+            status = os.unpackBoolean() ? LRAStatus.valueOf(os.unpackString()) : LRAStatus.Active;
```
`os.unpackBoolean()` will be always true

ochaloup replied:
+1

xstefank commented on rts/lra/lra-coordinator-jar/src/main/java/io/narayana/lra/coordinator/domain/model/Transaction.java:
```diff
@@ -362,24 +371,27 @@ protected void setLRAStatus(int actionStatus) {
     boolean isClosed() {
-        return status != null && status.equals(LRAStatus.Closed);
+        assert status != null;
+        return status.equals(LRAStatus.Closed);
```
you gave me an idea earlier in the PR :) maybe we can rewrite these comparisons to use `==` as they are enums. We can save some processing.

mmusgrov replied:
+1

Also `isClosed`, `isCancelled`, `isActive` etc are not required since any caller just needs to call `getStatus()` and then do the test (in fact I think they are unused). Also note that `class Transaction` is internal to the implementation so we have full control over the users of this class.
 
The other methods such as `isFinished` and `isRecovering` are useful since the check multiple states.

xstefank commented on rts/lra/lra-coordinator-jar/src/main/java/io/narayana/lra/coordinator/domain/service/LRAService.java:
```diff
@@ -124,45 +123,41 @@ public synchronized ReentrantLock tryLockTransaction(URI lraId) {
     }
 
-    public List<LRAStatusHolder> getAllRecovering() {
+    public List<LRAData> getAllRecovering() {
```
we should use unified either Collection or List but I don't see the reason to mix.

mmusgrov replied:
+1
I would choose `List` since there is a natural ordering for LRAs.

ochaloup replied:
+1, good point, I was thinking to use the `Collection` on `getAll` method for I can use the `Set` to prevent duplication of the records in the result collection. Then it happens I introduced this ambiguity. But I can switch back to `List`.

mmusgrov commented on rts/lra/lra-coordinator-jar/src/main/java/io/narayana/lra/coordinator/domain/model/Transaction.java:
```diff
@@ -145,12 +157,8 @@ public boolean save_state(OutputObjectState os, int ot) {
-                os.packString(status.name());
-            }
+            os.packBoolean(true);
+            os.packString(status == null ? LRAStatus.Active.name() : status.name());
```
status can never be null.

mmusgrov replied:
the null check is superfluous

ochaloup replied:
+1

mmusgrov commented on rts/lra/lra-coordinator-jar/src/main/java/io/narayana/lra/coordinator/domain/model/Transaction.java:
```diff
@@ -351,9 +359,10 @@ protected LRAService getLraService() {
      * return the current status of the LRA
      *
-     * @return null if the LRA is still active (not closing, cancelling, closed or cancelled
+     * @return current LRA status, will not return <code>null</code>
```
The text after the comma is not need (and therefore confusing).
status can never be null.

mmusgrov commented on rts/lra/lra-coordinator-jar/src/main/java/io/narayana/lra/coordinator/domain/model/Transaction.java:
```diff
@@ -351,9 +359,10 @@ protected LRAService getLraService() {
+     * @return current LRA status, will not return <code>null</code>
      */
     public LRAStatus getLRAStatus() {
+        assert status != null;
```
Agreed, status can never be null. But why do we need to be asserting that.

ochaloup replied:
assertion is good to demonstrate that the code is designed in way where the `status` won't be null. The developer says that he is aware of that fact and announce it to anybody reading the code later.

mmusgrov replied:
If we assert program invariant's everywhere they exist then the code would be littered with assertions making the program difficult to read.

Add a comment to the field if you really want to stress that the status can only take values according to a well defined state model

ochaloup replied:
agree, I think these are superfluous. I will change this.

mmusgrov commented on rts/lra/lra-coordinator-jar/src/main/java/io/narayana/lra/coordinator/domain/service/LRAService.java:
```diff
@@ -474,4 +463,9 @@ void disableRecovery() {
         }
     }
+
+    private Collection<LRAData> getLraDataByStatus(Map<URI, Transaction> lrasToFilter, LRAStatus status) {
```
Do we need to include `Lra` in the method names, it is clear that these methods are collecting LRAs.
But using it in `lrasToFilter` is good since it explains that the map contains LRAs.

ochaloup replied:
+1,  `getDataByStatus` is fine

mmusgrov commented on rts/lra/lra-coordinator-jar/src/main/java/io/narayana/lra/coordinator/internal/LRARecoveryModule.java:
```diff
@@ -195,7 +195,8 @@ public void getRecoveringLRAs(Map<URI, Transaction> lras) {
      * remove an LRA log record
-     * @param lraUid
+     *
+     * @param lraUid lra id that will be removed from the log record
```
LRA is an acronym so it should be capitalised (ie lra -> LRA) but obviously this comment does not apply to java variable names.

ochaloup replied:
+1

ochaloup commented on rts/lra/lra-coordinator-jar/src/main/java/io/narayana/lra/coordinator/domain/model/LRAData.java:
```diff
@@ -97,18 +83,102 @@ public int hashCode() {
+        }
+
+        public static BuilderWithLraId instanceOf() {
+            return new Builder();
```
I want not to require the order of the methods particularly but it's a secondary effect.
What I want is to require that the all the arguments are defined. For the `build()` method could be called and the object is created the developer has to load all data through the builder.

mmusgrov replied:
This is still complicated and difficult for me to understand, I would not like to have to maintain this code.
This is internal class and is for our use only - only the JSON representation is ever exposed (ie it isn't part of any external API).

ochaloup commented on rts/lra/lra-coordinator-jar/src/main/java/io/narayana/lra/coordinator/domain/model/LRAData.java:
```diff
@@ -97,18 +83,102 @@ public int hashCode() {
+            // no public instance
+        }
+
+        public static BuilderWithLraId instanceOf() {
```
+1

ochaloup commented on rts/lra/lra-coordinator-jar/src/main/java/io/narayana/lra/coordinator/domain/model/LRAData.java:
```diff
@@ -72,12 +58,12 @@ public long getFinishTime() {
-    public ZoneOffset getZoneOffset() {
-        return ZoneOffset.UTC;
+    public String getResponseData() {
+        return this.responseData;
```
+1

ochaloup replied:
I removed the `responseData` from `LRAData` while I left them in `LRARecord` where it seems they are used.

ochaloup commented on rts/lra/lra-coordinator-jar/src/main/java/io/narayana/lra/coordinator/api/Coordinator.java:
```diff
@@ -112,25 +110,23 @@
             @QueryParam(STATUS_PARAM_NAME) @DefaultValue("") String state) {
-        List<LRAStatusHolder> lras = lraService.getAll(state);
+        LRAStatus requestedLRAStatus = null;
+        if(state != null && !state.isEmpty()) {
```
+1

ochaloup commented on rts/lra/lra-coordinator-jar/src/main/java/io/narayana/lra/coordinator/domain/model/LRAData.java:
```diff
@@ -97,18 +83,102 @@ public int hashCode() {
+                isTopLevel, isRecovering, startTime, finishTime, httpStatus, responseData);
+    }
+
+    public static class Builder implements BuilderWithLraId, BuilderWithClientId, BuilderWithStatus, BuilderWithTopLevel,
```
I commented that above. I think it's easier to use a builder than a constructor with 8 parameters. This makes all the parameters required.

mmusgrov replied:
See my other comment on the complexity of the builder code.

I have a reasonable degree of experience with Java but I still don't really understand this code.

Constructors for DTOs are similar to copy constructors which conventionally take values for all those fields that cannot be calculated.

ochaloup replied:
I don't consider the builder code complex. It's just a Java code. It's a way how Java permits to build the API when it does not permit optional parameters or named ones.
I agree on the point that this is a private class and thus strictly with not a direct benefit.
I will change it for constructor.

ochaloup commented:
@mmusgrov @xstefank I hope I addressed your remarks. Please take a look once again now.

The only remaining review note is the  constructor vs. builder discussion. My personal opinion is that the builder is beneficial but if you both think that it makes the API misleading I will remove it. If you may, let me know.

ochaloup commented:
@mmusgrov I updated the PR
